salt/data-pipeline/bigquery-views.sls

```diff
@@ -0,0 +1,39 @@
+clone biquery-views repo:
+    builder.git_latest:
+        - name: git@github.com:elifesciences/data-pipeline-bigquery-views
+        - identity: {{ pillar.elife.projects_builder.key or '' }}
+        - rev: {{ salt['elife.rev']() }}
+        - branch: {{ salt['elife.branch']() }}
+        - target: /opt/data-pipeline-bigquery-views
+        - force_fetch: True
+        - force_checkout: True
+        - force_reset: True
+
+    cmd.run:
+        - cwd: /opt/data-pipeline-bigquery-views
+        - name: docker-compose build
```

Comment:
ah, docker support... then we should use the Docker image rather than rebuilding

Reply:
I'll add the publishing of the image to the repository so that this becomes straightforward :tada:

Reply:
The only thing is that the repo is not public and we decided that we rather not make it public now.

Reply:
It's acceptable to have private Docker images, doing open source is an exception. I'll compare Docker Hub vs. AWS container registry for hosting them.

Reply:
Docker Hub: `Purchase plans starting at $7/ month for 5 private repositories`
Good thing is that it is already configured in Docker daemons everywhere.

Reply:
Google's version is I believe just billed by the regular storage cost. Does AWS charge for the transfer even if you set it up for the same region? Charging by storage will encourage smaller images. In that case we would maybe just push the latest image rather than keeping the history.

Reply:
```
elifesciences/elife-xpub (~1GB)
Docker hub pull: 70s 
Docker hub push: 126s
AWS ECR pull: 71s                                                                                                                                                          
AWS ECR push: 125s
```
Pull time is mainly extraction rather than download, anyway. Given the comparable performance from an EC2 instance pulling and pushing, I'd go for Docker Hub to avoid lock in.

Reply:
There don't seem to be high costs involved for all these versions, as we incur them only for private projects (I think).

Reply:
> Charging by storage will encourage smaller images. 

no it won't. I point this out often, but there are no financial pressures at elife, at least, none that developers are exposed to. Being more expensive isn't going to stop us using a 4GB ubuntu-based container vs a 40MB Alpine one. The feedback between the dev's decision to go with a Ubuntu container and $4 extra dollars per month on the AWS bill just isn't there.

Just saying :) Happy Monday.

Reply:
> I'll add the publishing of the image to the repository

is this what causes the docker image to be built? 
https://github.com/elifesciences/data-pipeline-bigquery-views/blob/develop/Jenkinsfile#L5-L21

and will the publishing of the built image to Dockerhub live in the Jenkinsfile or elsewhere? I'm curious about how much logic lives in Alfred these days

Reply:
Yes it will be an explicit push.
Spending money won't stop big images, but waiting 5 minutes for each of your build to finish because it's pushing 4GB may.

Reply:
> waiting 5 minutes for each of your build to finish because it's pushing 4GB may.

definitely. 

I would *love* to tie how expensive a thing is to how long you have to wait for it ;) A multi-AZ RDS instance would then take half a day to bring up. 

It could also be argued that the developer time spent configuring and pushing a larger Ubuntu image is less expensive than developer time spent dealing with compiling things for the feature-less Alpine containers. For now I'm grateful that cost is not a dimension we optimise on, but I do get that urge sometimes to figure out how to run all of this in the cheapest way possible.